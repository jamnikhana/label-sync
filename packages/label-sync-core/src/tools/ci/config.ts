import { GithubRepository, getRepositoryFromName } from '../../github'
import { RepositoryConfig } from '../../types'
import { withDefault } from '../../utils'

export interface Config {
  [repository: string]: RepositoryConfig
}

export type ConfigError = { repository: string; message: string }

/**
 *
 * Hydrates repositories from configuration file.
 *
 * @param configuration
 */
export function getRepositoriesFromConfiguration(config: Config) {
  const repositories = Object.keys(config).reduce<{
    configurations: { repository: GithubRepository; config: RepositoryConfig }[]
    errors: ConfigError[]
  }>(
    (acc, repositoryName) => {
      const repository = getRepositoryFromName(repositoryName)

      if (!repository) {
        return {
          ...acc,
          errors: acc.errors.concat({
            repository: repositoryName,
            message: `Cannot decode the provided repository name ${repositoryName}`,
          }),
        }
      }

      return {
        ...acc,
        configurations: acc.configurations.concat({
          repository: repository,
          config: hydrateRepositoryConfig(config[repositoryName]),
        }),
      }
    },
    { configurations: [], errors: [] },
  )

  return repositories

  /**
   * Helpers
   */
  function hydrateRepositoryConfig(config: RepositoryConfig) {
    return {
      strict: withDefault(false)(config.strict),
      labels: config.labels,
    }
  }
}
