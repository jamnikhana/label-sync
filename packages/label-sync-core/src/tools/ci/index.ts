export { Config, getRepositoriesFromConfiguration } from './config'
export { handleSync } from './sync'
