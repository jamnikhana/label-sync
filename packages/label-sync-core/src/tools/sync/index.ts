import Octokit from '@octokit/rest'

import {
  Config,
  getRepositoriesFromConfiguration,
  RepositoryConfig,
} from '../../config'
import { GithubRepository } from '../../github'
import { handleLabelSync } from '../../handlers/labels'
import {
  SyncReport,
  RepositorySyncReport,
  RepositorySyncSuccessReport,
  RepositorySyncErrorReport,
  RepositorySyncConfigurationErrorReport,
} from './reporter'
import { handleSiblingSync } from '../../handlers/siblings'

export interface SyncOptions {
  dryRun: boolean
  githubToken: string
}

/**
 *
 * Handles repository sync by performing Label Sync first,
 * and Siblings Sync later.
 *
 * @param config
 * @param options
 */
export async function handleSync(
  config: Config,
  options: SyncOptions,
): Promise<SyncReport> {
  /**
   * Authentication
   */

  const client = new Octokit({
    headers: {
      accept: 'application/vnd.github.symmetra-preview+json',
    },
  })

  client.authenticate({
    type: 'app',
    token: options.githubToken,
  })

  /**
   * Repositories Sync
   */

  const repositories = getRepositoriesFromConfiguration(config)
  const repositoryHandlers = repositories.map(async repository => {
    if (repository.status === 'ok') {
      return handleRepository(repository.repository, repository.config)
    }

    return {
      status: 'config' as 'config',
      report: {
        config: repository.config,
        message: repository.message,
      },
    }
  })

  const repositoryReports = await Promise.all(repositoryHandlers)

  const report = generateSyncReport(repositoryReports)

  return {
    config: config,
    options: options,
    successes: report.successes,
    errors: report.errors,
    configs: report.configs,
  }

  /**
   * Helper functions
   */
  async function handleRepository(
    repository: GithubRepository,
    config: RepositoryConfig,
  ): Promise<RepositorySyncReport> {
    const labelSync = await handleLabelSync(client, repository, config, options)

    const siblingsSync = await handleSiblingSync(
      client,
      repository,
      config,
      options,
    )

    if (siblingsSync.status === 'error') {
      return {
        status: 'error',
        report: {
          repository: repository,
          config: config,
          message: siblingsSync.message,
        },
      }
    }

    return {
      status: 'success',
      report: {
        repository: repository,
        config: config,
        manifest: siblingsSync.manifest,
        additions: labelSync.additions,
        updates: labelSync.updates,
        removals: labelSync.removals,
        siblingsSucesses: siblingsSync.successes,
        siblingsErrors: siblingsSync.errors,
      },
    }
  }

  /**
   *
   * Merges multiple reports into one.
   *
   * @param repositoryReports
   */
  function generateSyncReport(
    repositoryReports: RepositorySyncReport[],
  ): {
    successes: RepositorySyncSuccessReport[]
    errors: RepositorySyncErrorReport[]
    configs: RepositorySyncConfigurationErrorReport[]
  } {
    const report = repositoryReports.reduce<{
      successes: RepositorySyncSuccessReport[]
      errors: RepositorySyncErrorReport[]
      configs: RepositorySyncConfigurationErrorReport[]
    }>(
      (acc, report) => {
        switch (report.status) {
          case 'success':
            return { ...acc, successes: [...acc.successes, report.report] }
          case 'error':
            return { ...acc, errors: [...acc.errors, report.report] }
          case 'config':
            return { ...acc, configs: [...acc.configs, report.report] }
        }
      },
      { successes: [], errors: [], configs: [] },
    )

    return report
  }
}
