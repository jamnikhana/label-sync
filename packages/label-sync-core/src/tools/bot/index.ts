import mls from 'multilines'
import Octokit from '@octokit/rest'
import * as probot from 'probot'

import { Config } from '../ci'
import { GithubIssue, GithubRepository } from '../../github'
import { RepositoryManifest, getRepositoryManifest } from '../../manifest'
import { assignSiblingsToIssue } from '../../handlers/siblings/siblings'
import { withDefault } from '../../utils'
import { getRepositoriesFromConfiguration } from '../ci/config'

export interface BotOptions {
  logger: Console
  githubToken: string
}

/**
 *
 * Creates a Github Bot instance based on provided configuration.
 *
 * @param config
 * @param logger
 */
export async function getGithubBot(
  config: Config,
  options: BotOptions,
): Promise<((app: probot.Application) => void)> {
  /* Options */

  const logger = withDefault(console)(options.logger)

  /* Authentication */

  const client = new Octokit({
    headers: {
      accept: 'application/vnd.github.symmetra-preview+json',
    },
  })

  client.authenticate({
    type: 'app',
    token: options.githubToken,
  })

  /* Generate manifest */

  const manifest = await generateBotManifest(client, config)

  /* Github Bot */

  const bot = (app: probot.Application) => {
    app.on('issues.labeled', async (context: probot.Context) => {
      const repository: GithubRepository = context.payload.repository
      const issue: GithubIssue = context.payload.issue

      /* Check repository configuration */
      if (!Object.keys(manifest).includes(repository.full_name)) {
        logger.log(`No such repository configuration, ${repository}.`)
        return
      }

      const res = await assignSiblingsToIssue(
        context.github,
        repository,
        issue,
        manifest[repository.full_name],
      )

      context.log.info(
        mls`
          | Successfully applied siblings to issue ${issue.number}: 
          | ${res}
          `,
      )
    })

    app.on('pull_request.labeled', async (context: probot.Context) => {
      const repository: GithubRepository = context.payload.repository
      const pr: GithubIssue = context.payload.pull_request

      /* Check repository configuration */
      if (!Object.keys(manifest).includes(repository.full_name)) {
        context.log.info(`No such repository configuration, ${repository}.`)
        return
      }

      const res = await assignSiblingsToIssue(
        context.github,
        repository,
        pr,
        manifest[repository.full_name],
      )

      context.log.info(
        mls`
          | Successfully applied siblings to ${pr.number} pull request: 
          | ${res}
          `,
      )
    })
  }

  return bot

  /* Helper functions */

  /**
   *
   * Generates manifests for each repo and combines them into one.
   *
   * @param config
   */
  async function generateBotManifest(
    github: Octokit,
    config: Config,
  ): Promise<{ [repository: string]: RepositoryManifest }> {
    const repositories = getRepositoriesFromConfiguration(config)

    if (repositories.errors.length !== 0) {
      throw new Error(`Couldn't generate manifest!`)
    }

    const manifests = await Promise.all(
      repositories.configurations.map(async repository => {
        const manifest = await getRepositoryManifest(
          github,
          repository.repository,
          repository.config,
        )

        return {
          repository: repository,
          manifest: manifest,
        }
      }),
    )

    const manifest = manifests.reduce((acc, manifest) => {
      if (manifest.manifest.status === 'err') {
        throw new Error(
          `Couldn't generate manifest: ${manifest.manifest.message}.`,
        )
      }

      return {
        ...acc,
        [manifest.repository.repository.full_name]: manifest.manifest.manifest,
      }
    }, {})

    return manifest
  }
}
