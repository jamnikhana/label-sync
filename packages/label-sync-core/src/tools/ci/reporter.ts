import chalk from 'chalk'
import ml from 'multilines'

import { GithubRepository } from '../../github'
import {
  LabelSyncReport,
  createTerminalReport as createLabelSyncReport,
} from '../../handlers/labels'
import {
  SiblingSyncReport,
  createTerminalReport as createSiblingSyncReport,
} from '../../handlers/siblings'
import { RepositoryManifest } from '../../manifest'
import { RepositoryConfig } from '../../types'

import { Config, ConfigError } from './config'
import { SyncOptions } from './sync'

export type SyncReport = {
  config: Config
  options: SyncOptions
  syncs: RepositorySyncReport[]
  configErrors: ConfigError[]
}

export type RepositorySyncReport =
  | {
      status: 'success'
      repository: GithubRepository
      config: RepositoryConfig
      manifest: RepositoryManifest
      labels: LabelSyncReport
      siblings: SiblingSyncReport | null
    }
  | {
      status: 'error'
      message: string
      repository: GithubRepository
      config: RepositoryConfig
    }

/**
 *
 * Creates a human readable terminal report of CI Sync.
 * (Uses chalk to make report more lively.)
 *
 * @param report
 */
export function createTerminalReport(report: SyncReport): string {
  return ml`
    | ${chalk.bgBlueBright('Label Sync Report')}
    | ${chalk.gray('This is an autogenerated report for your project.')}
    | (dry run: "${report.options.dryRun}")
    |
    | ${syncsList(report.syncs)}
    |
    | ${configErrorsList(report.configErrors)}
  `

  function syncsList(reports: RepositorySyncReport[]): string {
    return reports
      .map(report => {
        switch (report.status) {
          case 'success': {
            if (report.siblings) {
              return ml`
              | ${chalk.bgGreen(report.repository.full_name)}
              | ${createLabelSyncReport(report.labels)}
              |
              | ${createSiblingSyncReport(report.siblings)}
              `
            } else {
              return ml`
              | ${chalk.bgGreen(report.repository.full_name)}
              | ${createLabelSyncReport(report.labels)}
              `
            }
          }

          case 'error': {
            return ml`
              | ${chalk.bgRed(report.repository.full_name)}
              | ${report.message}
            `
          }
        }
      })
      .join('\n')
  }

  function configErrorsList(reports: ConfigError[]): string {
    if (reports.length === 0) return 'Synced all repositories with no problems!'
    return ml`
    | Check the configuration of these projects:
    | ${reports.map(report => report.message).join('\n')}
    `
  }
}
