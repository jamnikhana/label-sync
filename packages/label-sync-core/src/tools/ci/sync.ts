import Octokit from '@octokit/rest'

import { handleLabelSync } from '../../handlers/labels'
import { handleSiblingSync } from '../../handlers/siblings'
import { Config, getRepositoriesFromConfiguration } from './config'
import { SyncReport, RepositorySyncReport } from './reporter'
import { GithubRepository } from '../../github'
import { RepositoryConfig } from '../../types'
import { getRepositoryManifest } from '../../manifest'

export interface SyncOptions {
  dryRun: boolean
}

/**
 *
 * Handles repository sync by performing Label Sync first,
 * and Siblings Sync later.
 *
 * @param config
 * @param options
 */
export async function handleSync(
  client: Octokit,
  config: Config,
  options: SyncOptions,
): Promise<SyncReport> {
  const { configurations, errors } = getRepositoriesFromConfiguration(config)
  const syncs = await Promise.all(configurations.map(handleRepository))

  return {
    config: config,
    options: options,
    syncs: syncs,
    configErrors: errors,
  }

  /* Helper functions */
  async function handleRepository(configuration: {
    repository: GithubRepository
    config: RepositoryConfig
  }): Promise<RepositorySyncReport> {
    /**
     * Firstly, verify configuration. After that, sync labels and then
     * add missing siblings to existing issues.
     */

    const manifest = await getRepositoryManifest(
      client,
      configuration.repository,
      configuration.config,
    )

    /* Report manifest generation error */
    if (manifest.status !== 'ok') {
      return {
        status: 'error',
        message: `Error generating manifest: ${manifest.message}`,
        repository: configuration.repository,
        config: configuration.config,
      }
    }

    const labelSync = await handleLabelSync(
      client,
      configuration.repository,
      configuration.config,
      options,
    )

    const siblingsSync = await handleSiblingSync(
      client,
      configuration.repository,
      manifest.manifest,
      options,
    )

    return {
      status: 'success',
      repository: configuration.repository,
      config: configuration.config,
      manifest: manifest.manifest,
      labels: labelSync,
      siblings: siblingsSync,
    }
  }
}
