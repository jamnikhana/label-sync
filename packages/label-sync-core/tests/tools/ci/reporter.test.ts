import { SyncReport, createTerminalReport } from '../../../src/tools/ci'

describe('ci sync reporter', () => {
  test('correctly creates a report', async () => {
    const report: SyncReport = {
      config: {
        'maticzav/label-sync': { labels: { basic: '123123' }, strict: false },
        'maticzav/graphql-shield': {
          labels: {
            basic: '234234',
            'kind/bug': {
              color: 'ff0000',
              siblings: ['bug/0-no-reproduction'],
            },
            'bug/0-no-reproduction': {
              color: '00ff00',
              description: 'No reproduction available.',
            },
          },
          strict: false,
        },
      },
      options: { dryRun: false },
      syncs: [
        {
          status: 'success',
          repository: {
            owner: { login: 'maticzav' },
            name: 'label-sync',
            full_name: 'maticzav/label-sync',
          },
          config: { strict: false, labels: { basic: '123123' } },
          manifest: {
            basic: {
              label: {
                name: 'basic',
                color: 'f266f4',
                description: '',
                default: false,
              },
              siblings: [],
            },
            'bug/no-reproduction': {
              label: {
                name: 'bug/no-reproduction',
                color: '333333',
                description: '',
                default: false,
              },
              siblings: [],
            },
            'kind/bug': {
              label: {
                name: 'kind/bug',
                color: '333333',
                description: '',
                default: false,
              },
              siblings: [],
            },
          },
          labels: {
            repository: {
              owner: { login: 'maticzav' },
              name: 'label-sync',
              full_name: 'maticzav/label-sync',
            },
            config: { strict: false, labels: { basic: '123123' } },
            options: { dryRun: false },
            additions: [],
            updates: [
              {
                default: false,
                name: 'basic',
                description: '',
                color: '123123',
              },
            ],
            removals: [
              {
                name: 'bug/no-reproduction',
                color: '333333',
                description: '',
                default: false,
              },
              {
                name: 'kind/bug',
                color: '333333',
                description: '',
                default: false,
              },
            ],
          },
          siblings: {
            repository: {
              owner: { login: 'maticzav' },
              name: 'label-sync',
              full_name: 'maticzav/label-sync',
            },
            manifest: {
              basic: {
                label: {
                  name: 'basic',
                  color: 'f266f4',
                  description: '',
                  default: false,
                },
                siblings: [],
              },
              'bug/no-reproduction': {
                label: {
                  name: 'bug/no-reproduction',
                  color: '333333',
                  description: '',
                  default: false,
                },
                siblings: [],
              },
              'kind/bug': {
                label: {
                  name: 'kind/bug',
                  color: '333333',
                  description: '',
                  default: false,
                },
                siblings: [],
              },
            },
            options: { dryRun: false },
            issues: [
              {
                issue: {
                  id: 1421,
                  title: 'Issue Title',
                  body: 'Very serious issue.',
                  state: '',
                  labels: [
                    {
                      name: 'bug/no-reproduction',
                      color: '333333',
                      description: '',
                      default: false,
                    },
                    {
                      name: 'basic',
                      color: 'f266f4',
                      description: '',
                      default: false,
                    },
                    {
                      name: 'kind/bug',
                      color: '333333',
                      description: '',
                      default: false,
                    },
                  ],
                  number: 1,
                },
                siblings: [],
              },
              {
                issue: {
                  id: 1123,
                  title: 'Another Issue Title',
                  body: 'Not so serious issue.',
                  state: '',
                  labels: [
                    {
                      name: 'bug/no-reproduction',
                      color: '333333',
                      description: '',
                      default: false,
                    },
                    {
                      name: 'basic',
                      color: 'f266f4',
                      description: '',
                      default: false,
                    },
                    {
                      name: 'kind/bug',
                      color: '333333',
                      description: '',
                      default: false,
                    },
                  ],
                  number: 5,
                },
                siblings: [],
              },
            ],
          },
        },
        {
          config: {
            labels: {
              basic: '123123',
              complex: {
                color: '321321',
                siblings: ['doesntexist', 'kind/missing'],
              },
            },
            strict: true,
          },
          message:
            'Error generating manifest: Labels doesntexist, kind/missing are not defined',
          repository: {
            full_name: 'maticzav/label-sync',
            name: 'label-sync',
            owner: { login: 'maticzav' },
          },
          status: 'error',
        },
      ],
      configErrors: [
        {
          message: 'Cannot decode the provided repository name label-sync',
          repository: 'label-sync',
        },
      ],
    }

    expect(createTerminalReport(report)).toMatchSnapshot()
  })

  test('correctly creates an empty report', async () => {
    const report: SyncReport = {
      config: {
        'maticzav/label-sync': { labels: { basic: '123123' }, strict: false },
        'maticzav/graphql-shield': {
          labels: {
            basic: '234234',
            'kind/bug': {
              color: 'ff0000',
              siblings: ['bug/0-no-reproduction'],
            },
            'bug/0-no-reproduction': {
              color: '00ff00',
              description: 'No reproduction available.',
            },
          },
          strict: false,
        },
      },
      options: { dryRun: false },
      syncs: [],
      configErrors: [],
    }

    expect(createTerminalReport(report)).toMatchSnapshot()
  })
})
