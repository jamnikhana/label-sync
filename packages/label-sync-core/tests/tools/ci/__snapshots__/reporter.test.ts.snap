// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ci sync reporter correctly creates a report 1`] = `
"Label Sync Report
This is an autogenerated report for your project.
(dry run: "false")

maticzav/label-sync
This is an autogenerated report for your project.
(dry run: "false")

New Labels:
No new labels.

Updated Labels:
- basic

Removed Labels:
- bug/no-reproduction
- kind/bug

We haven't removed any label yet.
If you want us to remove them, set strict to true.

This is an autogenerated report for your project.
(dry run: "false")


maticzav/label-sync
Error generating manifest: Labels doesntexist, kind/missing are not defined

Check the configuration of these projects:
Cannot decode the provided repository name label-sync"
`;

exports[`ci sync reporter correctly creates a report when skipping siblings 1`] = `
"Label Sync Report
This is an autogenerated report for your project.
(dry run: "false")

maticzav/label-sync
This is an autogenerated report for your project.
(dry run: "false")

New Labels:
No new labels.

Updated Labels:
- basic

Removed Labels:
- bug/no-reproduction
- kind/bug

We haven't removed any label yet.
If you want us to remove them, set strict to true.
maticzav/label-sync
Error generating manifest: Labels doesntexist, kind/missing are not defined

Check the configuration of these projects:
Cannot decode the provided repository name label-sync"
`;

exports[`ci sync reporter correctly creates an empty report 1`] = `
"Label Sync Report
This is an autogenerated report for your project.
(dry run: "false")



Synced all repositories with no problems!"
`;

exports[`ci sync reporter correctly creates an empty report when skipping siblings 1`] = `
"Label Sync Report
This is an autogenerated report for your project.
(dry run: "false")



Synced all repositories with no problems!"
`;
